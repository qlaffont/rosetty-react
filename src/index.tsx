/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { createContext, useContext, useMemo, useState } from 'react';
import { Language, rosetty, RosettyReturn } from 'rosetty';

export const RosettyContext = createContext({});

export const RosettyProvider = ({
  children,
  languages,
  defaultLanguage,
  translateFallback,
}: {
  children: React.ReactNode;
  languages: Record<string, Language>;
  defaultLanguage: string;
  translateFallback?: boolean;
}) => {
  const r = useMemo(
    () => rosetty(languages, defaultLanguage, translateFallback),
    [languages]
  );
  const [actualLang, setActualLang] = useState(defaultLanguage);

  const providerReturn = useMemo(
    () => ({
      ...r,
      actualLang,
      changeLang: (lang: string) => {
        r.changeLang(lang);
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        setActualLang(r.getCurrentLang()!);
      },
    }),
    [actualLang, r]
  );

  return (
    <RosettyContext.Provider value={providerReturn}>
      {children}
    </RosettyContext.Provider>
  );
};

type AnyObject = Record<string, any>;

export function useRosetty<T extends AnyObject>(): RosettyReturn<T> & {
  actualLang: string | undefined;
} {
  return useContext(RosettyContext) as RosettyReturn<T> & {
    actualLang: string | undefined;
  };
}

export type Rosetty<T extends AnyObject> = RosettyReturn<T>;
