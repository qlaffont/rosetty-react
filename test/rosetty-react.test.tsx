/**
 * @jest-environment jsdom
 */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/ban-ts-comment */
// Importing the jest testing library
import '@testing-library/jest-dom';

import { cleanup, renderHook } from '@testing-library/react';
//@ts-ignore
import enGB from 'dayjs/locale/en-gb';
//@ts-ignore
import fr from 'dayjs/locale/fr';
import React from 'react';

// MUTE console error
// eslint-disable-next-line @typescript-eslint/no-empty-function
console.error = () => {};

import { RosettyContext, RosettyProvider, useRosetty } from '../src';
//@ts-ignore
const I18NContextProvider = ({ children, languages, defaultLanguage }) => (
  <RosettyProvider languages={languages} defaultLanguage={defaultLanguage}>
    {children}
  </RosettyProvider>
);

afterEach(() => {
  cleanup();
});

describe('rosetty react', () => {
  it('should export useI18n + HOC + context', () => {
    expect(typeof useRosetty).toBe('function');
    expect(typeof RosettyProvider).toBe('function');
    expect(typeof RosettyContext).toBe('object');
  });

  it('should be able to use i18n', () => {
    //@ts-ignore
    const wrapper = ({ children }) => (
      <I18NContextProvider
        languages={{ fr: { dict: {}, locale: fr } }}
        defaultLanguage="fr"
      >
        {children}
      </I18NContextProvider>
    );
    const { result } = renderHook(() => useRosetty(), { wrapper });

    expect(result.current.languages).toStrictEqual(['fr']);
    expect(Object.keys(result.current)).toStrictEqual([
      'changeLang',
      'languages',
      'getCurrentLang',
      't',
      'displayNames',
      'listFormat',
      'numberFormat',
      'pluralRules',
      'format',
      'formatRelative',
      'formatDistance',
      'formatDistanceToNow',
      'formatDuration',
      'actualLang',
    ]);
  });

  it('should be able to return error', () => {
    //@ts-ignore
    const wrongWrapperLanguageNotValid = ({ children }) => (
      <I18NContextProvider
        languages={{ fr: { dict: {}, locale: fr } }}
        defaultLanguage="en"
      >
        {children}
      </I18NContextProvider>
    );

    try {
      renderHook(() => useRosetty(), {
        wrapper: wrongWrapperLanguageNotValid,
      });
    } catch (error) {
      //@ts-ignore
      expect(error.message).toBe('rosetty: language en not found');
    }
  });

  it('should be able to return actualLang', () => {
    //@ts-ignore
    const wrapper = ({ children }) => (
      <I18NContextProvider
        languages={{
          fr: { dict: {}, locale: fr },
          en: { dict: {}, locale: enGB },
        }}
        defaultLanguage="en"
      >
        {children}
      </I18NContextProvider>
    );

    const { result, rerender } = renderHook(() => useRosetty(), { wrapper });

    expect(result.current.actualLang).toStrictEqual('en');
    result.current.changeLang('fr');
    rerender();
    expect(result.current.actualLang).toStrictEqual('fr');
  });
});
