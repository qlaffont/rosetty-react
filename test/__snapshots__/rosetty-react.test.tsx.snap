// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rosetty react should be able to use i18n 1`] = `
[
  "changeLang",
  "languages",
  "getCurrentLang",
  "t",
  "displayNames",
  "listFormat",
  "numberFormat",
  "pluralRules",
  "dateTimeFormat",
  "relativeTimeFormat",
  "collator",
  "segmenter",
  "durationFormat",
  "actualLang",
]
`;
